### PR TITLE
Use docker/distribution library used by docker 17.05.x to support all docker image reference.

### PR DESCRIPTION
**Pull request by Random-Liu**
We are using new `docker/distribution` instead of previous `docker/docker/reference` because the later will introduce too many unnecessary dependencies, and also cause runc version conflict.

With this PR, I could pull image now, but there are other problems still. cri-containerd panics during the cri validation image test, will fix them one by one.

```
Running Suite: E2ECRI Suite
===========================
Random Seed: 1494610089 - Will randomize all specs
Will run 6 of 36 specs

SSS
------------------------------
[k8s.io] Image Manager 
  public image with digest should be pulled and removed [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:60
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:48
[It] public image with digest should be pulled and removed [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:60
STEP: Pull image : busybox@sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
STEP: Check image list to make sure pulling image success : busybox@sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
STEP: Get image list for imageName : busybox@sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
STEP: Remove image : busybox@sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
STEP: Remove image : busybox@sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
STEP: Check image list empty
STEP: Get image list for imageName : busybox@sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
[AfterEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51

• Failure [1.864 seconds]
[k8s.io] Image Manager
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:72
  public image with digest should be pulled and removed [Conformance] [It]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:60

  Should have none image in list
  Expected
      <int>: 1
  to equal
      <int>: 0

  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:147
------------------------------
S
------------------------------
[k8s.io] Image Manager 
  image status get image fields should not be empty [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:71
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:48
[It] image status get image fields should not be empty [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:71
STEP: Pull image : busybox:1.26.2
STEP: Get image status
2017/05/12 10:28:12 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017/05/12 10:28:12 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017/05/12 10:28:12 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017/05/12 10:28:12 transport: http2Client.notifyError got notified that the client transport was broken EOF.
E0512 10:28:12.663198   73093 remote_image.go:77] ImageStatus "busybox:1.26.2" from image service failed: rpc error: code = 13 desc = transport is closing
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: write unix @->/var/run/cri-containerd.sock: write: broken pipe"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: write unix @->/var/run/cri-containerd.sock: write: broken pipe"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /var/run/cri-containerd.sock: connect: connection refused"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /var/run/cri-containerd.sock: connect: connection refused"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
May 12 10:28:12.663: INFO: Unexpected error occurred: rpc error: code = 13 desc = transport is closing
STEP: Remove image : busybox:1.26.2
E0512 10:28:12.664322   73093 remote_image.go:124] RemoveImage "busybox:1.26.2" from image service failed: rpc error: code = 14 desc = grpc: the connection is unavailable
May 12 10:28:12.664: INFO: Unexpected error occurred: rpc error: code = 14 desc = grpc: the connection is unavailable
[AfterEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51

• Failure [0.969 seconds]
[k8s.io] Image Manager
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:72
  image status get image fields should not be empty [Conformance] [It]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:71

  failed to get image status: rpc error: code = 13 desc = transport is closing
  Expected error:
      <*grpc.rpcError | 0xc4203d46e0>: {
          code: 13,
          desc: "transport is closing",
      }
      rpc error: code = 13 desc = transport is closing
  not to have occurred

  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:171
------------------------------
SSSS
------------------------------
[k8s.io] Image Manager 
  listImage should get exactly 3 repoTags in the result image [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:137
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:48
[It] listImage should get exactly 3 repoTags in the result image [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:137
STEP: Pull image : busybox:1.26.2-uclibc
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /var/run/cri-containerd.sock: connect: connection refused"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /var/run/cri-containerd.sock: connect: connection refused"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
E0512 10:28:12.667101   73093 remote_image.go:102] PullImage "busybox:1.26.2-uclibc" from image service failed: rpc error: code = 14 desc = grpc: the connection is unavailable
May 12 10:28:12.667: INFO: Unexpected error occurred: rpc error: code = 14 desc = grpc: the connection is unavailable
[AfterEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51

• Failure [0.002 seconds]
[k8s.io] Image Manager
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:72
  listImage should get exactly 3 repoTags in the result image [Conformance] [It]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:137

  failed to pull image: rpc error: code = 14 desc = grpc: the connection is unavailable
  Expected error:
      <*grpc.rpcError | 0xc420546d80>: {
          code: 14,
          desc: "grpc: the connection is unavailable",
      }
      rpc error: code = 14 desc = grpc: the connection is unavailable
  not to have occurred

  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:201
------------------------------
SSSSSSSS
------------------------------
[k8s.io] Image Manager 
  listImage should get exactly 3 image in the result list [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:103
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:48
[It] listImage should get exactly 3 image in the result list [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:103
STEP: Pull image : busybox:1-uclibc
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /var/run/cri-containerd.sock: connect: connection refused"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /var/run/cri-containerd.sock: connect: connection refused"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
E0512 10:28:12.669836   73093 remote_image.go:102] PullImage "busybox:1-uclibc" from image service failed: rpc error: code = 14 desc = grpc: the connection is unavailable
May 12 10:28:12.669: INFO: Unexpected error occurred: rpc error: code = 14 desc = grpc: the connection is unavailable
[AfterEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51

• Failure [0.002 seconds]
[k8s.io] Image Manager
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:72
  listImage should get exactly 3 image in the result list [Conformance] [It]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:103

  failed to pull image: rpc error: code = 14 desc = grpc: the connection is unavailable
  Expected error:
      <*grpc.rpcError | 0xc4203aa220>: {
          code: 14,
          desc: "grpc: the connection is unavailable",
      }
      rpc error: code = 14 desc = grpc: the connection is unavailable
  not to have occurred

  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:201
------------------------------
SSSS
------------------------------
[k8s.io] Image Manager 
  public image with tag should be pulled and removed [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:52
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:48
[It] public image with tag should be pulled and removed [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:52
STEP: Pull image : busybox:1.26.2
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /var/run/cri-containerd.sock: connect: connection refused"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /var/run/cri-containerd.sock: connect: connection refused"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
E0512 10:28:12.672385   73093 remote_image.go:102] PullImage "busybox:1.26.2" from image service failed: rpc error: code = 14 desc = grpc: the connection is unavailable
May 12 10:28:12.672: INFO: Unexpected error occurred: rpc error: code = 14 desc = grpc: the connection is unavailable
[AfterEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51

• Failure [0.001 seconds]
[k8s.io] Image Manager
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:72
  public image with tag should be pulled and removed [Conformance] [It]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:52

  failed to pull image: rpc error: code = 14 desc = grpc: the connection is unavailable
  Expected error:
      <*grpc.rpcError | 0xc42044c760>: {
          code: 14,
          desc: "grpc: the connection is unavailable",
      }
      rpc error: code = 14 desc = grpc: the connection is unavailable
  not to have occurred

  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:201
------------------------------
SSSSSSSSS
------------------------------
[k8s.io] Image Manager 
  public image without tag should be pulled and removed [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:56
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:48
[It] public image without tag should be pulled and removed [Conformance]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:56
STEP: Pull image : busybox:latest
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /var/run/cri-containerd.sock: connect: connection refused"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
2017/05/12 10:28:12 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /var/run/cri-containerd.sock: connect: connection refused"; Reconnecting to {/var/run/cri-containerd.sock <nil>}
E0512 10:28:12.674088   73093 remote_image.go:102] PullImage "busybox:latest" from image service failed: rpc error: code = 14 desc = grpc: the connection is unavailable
May 12 10:28:12.674: INFO: Unexpected error occurred: rpc error: code = 14 desc = grpc: the connection is unavailable
[AfterEach] [k8s.io] Image Manager
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51

• Failure [0.002 seconds]
[k8s.io] Image Manager
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:72
  public image without tag should be pulled and removed [Conformance] [It]
  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:56

  failed to pull image: rpc error: code = 14 desc = grpc: the connection is unavailable
  Expected error:
      <*grpc.rpcError | 0xc4203aab20>: {
          code: 14,
          desc: "grpc: the connection is unavailable",
      }
      rpc error: code = 14 desc = grpc: the connection is unavailable
  not to have occurred

  /usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:201
------------------------------
S

Summarizing 6 Failures:

[Fail] [k8s.io] Image Manager [It] public image with digest should be pulled and removed [Conformance] 
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:147

[Fail] [k8s.io] Image Manager [It] image status get image fields should not be empty [Conformance] 
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:171

[Fail] [k8s.io] Image Manager [It] listImage should get exactly 3 repoTags in the result image [Conformance] 
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:201

[Fail] [k8s.io] Image Manager [It] listImage should get exactly 3 image in the result list [Conformance] 
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:201

[Fail] [k8s.io] Image Manager [It] public image with tag should be pulled and removed [Conformance] 
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:201

[Fail] [k8s.io] Image Manager [It] public image without tag should be pulled and removed [Conformance] 
/usr/local/google/home/lantaol/workspace/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/image.go:201

Ran 6 of 36 Specs in 2.843 seconds
FAIL! -- 0 Passed | 6 Failed | 0 Pending | 30 Skipped --- FAIL: TestE2ECRI (2.84s)
FAIL

Ginkgo ran 1 suite in 3.125705817s
Test Suite Failed
exit status 1
```

### CODE REVIEW

File: pkg/server/image_pull.go [hunk 1]
```diff
@@ -66,6 +77,16 @@ func (c *criContainerdService) PullImage(ctx context.Context, r *runtime.PullIma
 	return &runtime.PullImageResponse{ImageRef: digest}, err
 }
 
+// normalizeImageRef normalizes the image reference following the docker convention. This is added
+// mainly for backward compatibility.
```
Comment (mikebrow) [inline pkg/server/image_pull.go]: TODO unit test

File: pkg/server/image_pull.go [hunk 2]
```diff
@@ -20,28 +20,39 @@ import (
...
-		err  error
 		size int64
 		desc imagespec.Descriptor
 	)
 
-	image := r.GetImage().Image
+	image, err := normalizeImageRef(r.GetImage().GetImage())
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse image reference %q: %v", r.GetImage().GetImage(), err)
+	}
```
Comment (mikebrow) [inline pkg/server/image_pull.go]:
```
if r.GetImage().GetImage() != image {
	glog.V(2).Infof("PullImage using normalized ref: %q with auth config %+v", image)
}
```

File: pkg/server/image_pull.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/server/image_pull.go]: It's purely docker function. I'm not sure whether we should test it.

File: pkg/server/image_pull.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/server/image_pull.go]: Done.

**Comment by Random-Liu** [comment]
@mikebrow Addressed comments.

File: pkg/server/image_pull.go (hunk 1 above)
Comment (mikebrow) [inline pkg/server/image_pull.go]: yeah but if they change it we're letting them break customers.. sort of like when they changed their version tag to non sem-ver..

Comment (Random-Liu) [inline pkg/server/image_pull.go]: Done.

**Comment by Random-Liu** [comment]
Will merge the sandbox PR first, because this one is easier to rebase.

**Comment by Random-Liu** [comment]
Apply LGTM based on https://github.com/kubernetes-incubator/cri-containerd/pull/34#pullrequestreview-37930229.

**Pull request merged**